editor/src/components/canvas/controls/elements-outside-visible-area-hooks.tsx

```diff
@@ -0,0 +1,308 @@
+        x: topLeftPoint.x,
+        y: topLeftPoint.y,
+        width: frame.width * canvasScale,
+        height: frame.height * canvasScale,
+      } as WindowRectangle
+
+      const directions = getOutsideDirections(scaledCanvasArea, elementRect)
+      if (directions.length === 0) {
+        return null
+      }
+
+      return {
+        path: path,
+        rect: elementRect,
+        directions: directions,
+      }
+    }, elements)
+  }, [elements, canvasOffset, canvasScale, scaledCanvasArea, framesByPathString, navigatorWidth])
+
+  return React.useMemo((): ElementOutsideVisibleAreaIndicator[] => {
+    if (
+      scaledCanvasArea == null ||
+      scaledCanvasAreaCenter == null ||
+      bounds == null ||
+      canvasToolbar == null
+    ) {
+      return []
+    }
+
+    return (
+      elementsOutsideVisibleArea
+        // Map elements to indicators
+        .map(({ rect, path, directions }): ElementOutsideVisibleAreaIndicator => {
+          return {
+            id: getIndicatorId(path, directions),
+            path: path,
+            cluster: 1,
+            angle: angleBetweenPoints(scaledCanvasAreaCenter, getRectCenter(rect)),
+            position: adjustPosition(
+              offsetPoint(rect, {
+                x: rect.width / 2,
+                y: rect.height / 2,
+              } as WindowPoint),
+              directions,
+              scaledCanvasArea,
+              navigatorWidth,
+              windowRectangleFromDOMRect(canvasToolbar),
+            ),
+          }
+        })
+        // Group the indicators into clusters
+        .reduce((arr, indicator): ElementOutsideVisibleAreaIndicator[] => {
+          const index = arr.findIndex((other) => {
+            const distanceBetween = distance(indicator.position, other.position)
+            return distanceBetween < minClusterDistance
+          })
+          if (index >= 0) {
+            arr[index].cluster++
+          } else {
+            arr.push(indicator)
+          }
+          return arr
+        }, [] as ElementOutsideVisibleAreaIndicator[])
+    )
+  }, [
+    elementsOutsideVisibleArea,
+    scaledCanvasArea,
+    scaledCanvasAreaCenter,
+    navigatorWidth,
+    bounds,
+    canvasToolbar,
+  ])
+}
+
+function angleBetweenPoints(from: WindowPoint, to: WindowPoint): number {
+  return Math.atan2(to.y - from.y, to.x - from.x) + Math.PI
+}
+
+type ElementOutsideVisibleAreaDirectionBaseValue = {
+  direction: ElementOutsideVisibleAreaDirection
+  baseValue: number
+}
+
+function getPositionAxisRelativeToDirection(
+  directions: ElementOutsideVisibleAreaDirection[],
+  currentValue: number,
+  min: ElementOutsideVisibleAreaDirectionBaseValue,
+  max: ElementOutsideVisibleAreaDirectionBaseValue,
+): number {
+  if (directions.includes(min.direction)) {
+    return min.baseValue
+  } else if (directions.includes(max.direction)) {
+    return max.baseValue
+  } else {
+    return Math.max(Math.min(currentValue, max.baseValue), min.baseValue)
```

Comment:
We've got a function `clamp` for this BTW.

Reply:
noice! https://github.com/concrete-utopia/utopia/pull/3842/commits/a10d54815cd6720d040cc5570f6057d7f7c04107